lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -2535,6 +2544,281 @@ void StaticRepeatOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
+    ret = operation->getResult(2);
+  } else if (isa<InstanceOp>(operation)) {
+    // Get the go port of the instance through the "go" attribute
+    InstanceOp instanceOp = cast<InstanceOp>(operation);
+    auto portInfo = instanceOp.getReferencedComponent().getPortInfo();
+    for (auto [portInfo, res] : llvm::zip(portInfo, operation->getResults())) {
+      if (portInfo.hasAttribute("go"))
+        ret = res;
+    }
+  } else if (isa<PrimitiveOp>(operation)) {
+    // Get the go part of the primitive through the "calyx.go" attribute
+    PrimitiveOp primOp = cast<PrimitiveOp>(operation);
+    auto moduleExternOp = primOp.getReferencedPrimitive();
+    auto argAttrs = moduleExternOp.getArgAttrsAttr();
+    for (auto [attr, res] : llvm::zip(argAttrs, primOp.getResults())) {
+      if (DictionaryAttr dictAttr = dyn_cast<DictionaryAttr>(attr)) {
+        if (!dictAttr.empty()) {
+          if (dictAttr.begin()->getName().getValue() == "calyx.go")
+            ret = res;
+        }
+      }
+    }
+  }
+  return ret;
+}
+
+// Get the done port of the invoked component.
+Value InvokeOp::getInstDoneValue() {
+  ComponentOp componentOp = (*this)->getParentOfType<ComponentOp>();
+  Operation *operation = componentOp.lookupSymbol(getCallee());
+  Value ret = nullptr;
+  // The done port of these instances is the last result, so the ssa value
+  // of the last result is taken.
+  if (isa<RegisterOp, MemoryOp, DivSPipeLibOp, DivUPipeLibOp, MultPipeLibOp,
+          RemSPipeLibOp, RemUPipeLibOp>(operation)) {
+    size_t doneIdx = operation->getResults().size() - 1;
+    ret = operation->getResult(doneIdx);
+  } else if (isa<InstanceOp>(operation)) {
+    //  Get the go port of the instance through the "done" attribute
+    InstanceOp instanceOp = cast<InstanceOp>(operation);
+    auto portInfo = instanceOp.getReferencedComponent().getPortInfo();
+    for (auto [portInfo, res] : llvm::zip(portInfo, operation->getResults())) {
+      if (portInfo.hasAttribute("done"))
+        ret = res;
+    }
+  } else if (isa<calyx::PrimitiveOp>(operation)) {
+    // Get the go part of the primitive through the "calyx.done" attribute
+    PrimitiveOp primOp = cast<PrimitiveOp>(operation);
+    auto moduleExternOp = primOp.getReferencedPrimitive();
+    auto resAttrs = moduleExternOp.getResAttrsAttr();
+    for (auto [attr, res] : llvm::zip(resAttrs, primOp.getResults())) {
+      if (DictionaryAttr dictAttr = dyn_cast<DictionaryAttr>(attr)) {
+        if (!dictAttr.empty()) {
+          if (dictAttr.begin()->getName().getValue() == "calyx.done")
+            ret = res;
+        }
+      }
+    }
+  }
+  return ret;
+}
+
+// A helper function that gets the number of go or done ports in
+// hw.module.extern.
+static size_t
+getHwModuleExtGoOrDonePortNumber(hw::HWModuleExternOp &moduleExternOp,
+                                 bool isGo) {
+  size_t ret = 0;
+  std::string str = isGo ? "calyx.go" : "calyx.done";
+  for (Attribute attr : moduleExternOp.getArgAttrsAttr())
```

Comment:
This is a case where you'd be expected to add brackets for readability's sake (https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements).